# Intelligent Systems

## Assignment 2. Programming Intelligent Agents
- Eider Diaz A00828174, Campus Monterrey
- Carlos Hinojosa A01137566, Campus Monterrey
- Miguel Cortes A01270966, Campus Monterrey


# SIMPLE-REFLEX AGENT in FULLY OBSERVABLE ENVIRONMENT #

The grid enviroment composed of 6 rows and 6 columns is represented by an "island" with different objects placed randomly placed in some cells of the grid. As it was stated in the assigment, the location of all Thing objects in the environment is described by their \[r,c\] coordinates, where r represents rows and c is for columns.

The environment may contain 5 types of Thing Objects:
<ul>
    <li>Treasure 1 (T)</li>
    <li>Treasure 2 (t)</li>
    <li>Disposable Tool (h)</li>
    <li>Reusable Tool (H)</li>
    <li>Wall (X)</li>
</ul>

<li><b>percept(self, agent)</b> in this case outputs the contents of the whole grid environment in each step of a run. This means the agent is able to percieve or sense the entirety of the environment at any time. We programmed the percept as a list of objects belonging to the different classes of things that can be placed in the environment. This is convinient because any property of such objects may be accessed from that single percept list (most notably the .location of each object). </li>
<li><b>ReflexHunter</b> is the class of our simple-reflex agent. The main characteristic of this type of agent is that it lacks an "agent.memories" array to store useful objects it has seen in the environment in previous percepts of previous steps of a run. This type of agent can only output actions based on the current percept of the current step. We theorize that for a fully observable environment this won't make much of a difference since the full environment is always available as a single percept anyways.</li> 

The simple-reflex agent can move inside the island in four directions “MoveRight”, “MoveLeft”, “MoveUp”, and “MoveDown”. We also included the action "MoveRandom" for when an agent can't decide where to move to based on their percepts. This is specially useful to allow a simple reflex agent to be stuck in an unlucky situation. 

The agent also has actions for grabbing Things as described on the assignment. When certain conditions are met, the agent is able to perform the actions "greuse", "gdispos", "gTreasure1" and "gTreasure2" to grab a reusable tool, a disposable tool, a type 1 treasure and a type 2 treasure, respectively.

Some auxiliary methods on the program include "printMatrix" to show the current state of the Island, "inInventory" for the agent to check if they have the required tool to grab a treasure and "getDirection", for the agent to decide where to move to, based on the list of percepts. 

The program that controls how the agent acts upon it's environment is shown in the table below.

<table>
    <tbody><tr>
        <td><b>Percept:</b> </td>
        <td>Find tool (H) </td>
        <td>Find tool (h)</td>
        <td>Find treasure (T)</td>
         <td>Find treasure (t)</td>
      

   </tr>
   <tr>
       <td><b>Action (if not something grabbable in agent´s location move towards an useful object):</b> </td>
       <td>pick it</td>
       <td>pick it</td>
       <td>if has tool (H or h), Open it</td>
       <td>if has tool (h), Open it</td>
       
       
   </tr>
        
</tbody></table>

The program is done when it runs out of user-defined iterations or when there are no more Things on the Island, including Treasure1, Treasure2, Disposable Tool and Reusable Tool. 


<b>Next cell runs a simple reflex agent program in a fully observable environment. Output depicts the behavior of the agent for all the steps in a 50 iteration run with 3 items of each kind placed randomly in the environment. The starting position of the agent for this run is \[1, 1\]<b></b></b>

In [30]:
from agents import *
from random import *


class Treasure1(Thing):
    pass


class Treasure2(Thing):
    pass


class DisposTool(Thing):
    pass


class ReuseTool(Thing):
    pass


class Island(Environment):
    def __init__(self, width=7, height=7):
        super(Island, self).__init__()

        self.width = width
        self.height = height
        # Sets iteration start and end (no walls).
        # self.x_start, self.y_start = (0, 0)
        # self.x_end, self.y_end = (self.width, self.height)

    def percept(self, agent):
        '''prints & return a list of things that are in our agent's location'''
        percepts = []
        locations = []
        # locations = self.getLocations(agent.location)
        # TODO: implement 'getLocations function
        for i in range(1,7):
            for j in range(1,7):
                locations.append([i,j])

        for location in locations:
            things = self.list_things_at(location)
            for thing in things:
                percepts.append(thing)
        return percepts

    def execute_action(self, agent, action):
        '''changes the state of the environment based on what the agent does.'''
        # TODO: check for walls and boundaries
        if action == 'moveRandom':
            direction = randint(1, 4)
            print("Hunter: Moved random")
            if direction == 1:
                action = 'moveRight'
            elif direction == 2:
                action = 'moveLeft'
            elif direction == 3:
                action = 'moveUp'
            elif direction == 4:
                action = 'moveDown'

        if action == 'moveRight':
            if agent.location[1] < 6:
                walls = self.list_things_at([agent.location[0], agent.location[1]+1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveRight()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == 'moveLeft':
            if agent.location[1] > 1:
                walls = self.list_things_at([agent.location[0], agent.location[1] - 1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveLeft()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == 'moveUp':
            if agent.location[0] > 1:
                walls = self.list_things_at([agent.location[0]-1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveUp()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == 'moveDown':
            if agent.location[0] < 6:
                walls = self.list_things_at([agent.location[0] + 1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveDown()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == "Greuse":
            items = self.list_things_at(agent.location, tclass=ReuseTool)
            if len(items) != 0:
                if agent.greuse(items[0]):  #
                    self.delete_thing(items[0])  #
                    agent.holding.append('H')
        elif action == "Gdispos":
            agent.gdispos()
            items = self.list_things_at(agent.location, tclass=DisposTool)
            if len(items) != 0:
                if agent.gdispos(items[0]):  #
                    self.delete_thing(items[0])  # D
                    agent.holding.append('h')
        elif action == "GTreasure1":
            items = self.list_things_at(agent.location, tclass=Treasure1)
            if len(items) != 0:
                if agent.gTreasure1(items[0]):  # Grab Treasure 1
                    # TODO: add to performance
                    agent.performance += 20
                    self.delete_thing(items[0])  # Delete it from the Island after.
        elif action == "GTreasure2":
            items = self.list_things_at(agent.location, tclass=Treasure2)
            if len(items) != 0:
                if agent.gTreasure2(items[0]):  # Grab Treasure2
                    agent.performance += 40
                    agent.holding.remove('h')
                    self.delete_thing(items[0])  # Delete it from the Island after.
        elif action == "NoOp":
            pass

    def is_done(self):
        '''By default, we're done when we can't find a live agent,
        but to prevent killing our cute dog, we will or it with when there is no more food or water'''
        no_edibles = not any(isinstance(thing, Treasure1) or isinstance(thing, DisposTool) or isinstance(thing, ReuseTool) or isinstance(thing, Treasure2) for thing in self.things)
        dead_agents = not any(agent.is_alive() for agent in self.agents)
        return dead_agents or no_edibles

    def step(self): #AIMA function overriden to access agent status in each step of a run and print the performance
        """Run the environment for one time step. If the
        actions and exogenous changes are independent, this method will
        do.  If there are interactions between them, you'll need to
        override this method."""
        if not self.is_done():
            actions = []
            for agent in self.agents:
                if agent.alive:
                    actions.append(agent.program(self.percept(agent), agent))
                else:
                    actions.append("")
                
            for (agent, action) in zip(self.agents, actions):
                self.execute_action(agent, action)
            self.exogenous_change()

class ReflexHunter(Agent):
    def __init__(self, program=None):  # default state of the agent
        self.alive = True
        self.bump = False
        self.type = 'simple_reflex'
        self.holding = []
        self.performance = 50
        if program is None:
            def program(percept):
                return eval(input('Percept={}; action? '.format(percept)))
        # assert isinstance(program, collections.Callable)
        self.program = program

    def moveRight(self):
        self.location[1] += 1
        print("Hunter: Moved Right to {}.".format(self.location))

    def moveLeft(self):
        self.location[1] -= 1
        print("Hunter: Moved Left to {}.".format(self.location))

    def moveUp(self):
        self.location[0] -= 1
        print("Hunter: Moved Up to {}.".format(self.location))

    def moveDown(self):
        self.location[0] += 1
        print("Hunter: Moved Down to {}.".format(self.location))

    def greuse(self, thing):
        '''returns True upon success or False otherwise'''
        if isinstance(thing, ReuseTool):
            print("Hunter: Grabbed Reusable Tool at {}.".format(self.location))
            return True
        return False

    def gdispos(self, thing = None):
        ''' returns True upon success or False otherwise'''
        if isinstance(thing, DisposTool):
            print("Hunter: Grabbed Disposable Tool at {}.".format(self.location))
            return True
        return False

    def gTreasure1(self, thing):
        ''' returns True upon success or False otherwise'''
        if isinstance(thing, Treasure1):
            print("Hunter: Grabbed Treasure1 at {}.".format(self.location))
            return True
        return False

    def gTreasure2(self, thing):
        ''' returns True upon success or False otherwise'''
        if isinstance(thing, Treasure2):
            print("Hunter: Grabbed Treasure2 at {}.".format(self.location))
            return True
        return False
    
def program(percepts, agent):
    '''Returns an action based on it's percepts'''
    print("Agent Location: " + str(agent.location))
    print("Agent Tools: " + str(agent.holding))
    printMatrix(percepts)
    print("")
    print("Agent performance: " + str(agent.performance))
    if agent.type == 'modelbased':
        print("INTERNAL STATE:")
        print(agent.seen)
    actionTaken = False
    for p in percepts:
        # Grab actions for when agent is in same location
        if actionTaken:
            break
        in_location = agent.location == p.location
        if isinstance(p, Treasure1) and inInventory('H', agent):
            if in_location:
                actionTaken = True
                return 'GTreasure1'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
        elif isinstance(p, Treasure2) and inInventory('h', agent):
            if in_location:
                actionTaken = True
                return 'GTreasure2'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
        elif isinstance(p, DisposTool):
            if in_location:
                actionTaken = True
                return 'Gdispos'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
        elif isinstance(p, ReuseTool):
            if in_location:
                actionTaken = True
                return 'Greuse'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo

    if not actionTaken:
        return 'moveRandom'

def printMatrix(percepts):
    for i in range(0, 8):
        print("")
        for j in range(0, 8):
            if (i == 0 and j == 0) or (i == 7 and j == 7):
                print("\\", end=" ")

            elif (i == 0 and j == 7) or (i == 7 and j == 0):
                print("/", end=" ")

            elif (i == 0 or i == 7):
                print("".join(str(j)), end=" ")

            elif (j == 0 or j == 7):
                print("".join(str(i)), end=" ")

            else:
                #Check percepts, print if matches location
                printed = False
                for p in percepts:
                    if p.location == [i,j]:
                        if isinstance(p, Treasure1):
                            printed = True
                            print("T", end=" ")
                        elif isinstance(p, Treasure2):
                            printed = True
                            print("t", end=" ")
                        elif isinstance(p, DisposTool):
                            printed = True
                            print("h", end=" ")
                        elif isinstance(p, ReuseTool):
                            printed = True
                            print("H", end=" ")
                        elif isinstance(p, Wall):
                            printed = True
                            print("X", end=" ")
                if not printed:
                    print("-", end=" ")


def inInventory(tool, agent):
    for hold in agent.holding:
        if hold == tool:
            return True
    return False

def getDirection(origin, goal):
    if origin[1] < goal[1]:
        return 'moveRight'
    elif origin[1] > goal[1]:
        return 'moveLeft'
    elif origin[0] > goal[0]:
        return 'moveUp'
    else:
        return 'moveDown'
    
def main(start = [1, 1]):
    '''Run the simple reflex hunter in the partial treasure island environment'''
    island = Island()
    charlie = ReflexHunter(program)

    #Create 3 objects of each kind to place on the environment
    treasure1A = Treasure1()
    treasure1B = Treasure1()
    treasure1C = Treasure1()
    treasure2A = Treasure2()
    treasure2B = Treasure2()
    treasure2C = Treasure2()
    dispos1 = DisposTool()
    dispos2 = DisposTool()
    dispos3 = DisposTool()
    reusable1 = ReuseTool()
    reusable2 = ReuseTool()
    reusable3 = ReuseTool()
    wall = Wall()
    wall2 = Wall()
    wall3 = Wall()
    
    #add things randomly preventing placing more than one item in a single cell in the environment
    xy = [[1, 1],[1, 2],[1, 3],[1, 4],[1, 5],[1, 6],
          [2, 1],[2, 2],[2, 3],[2, 4],[2, 5],[2, 6],
          [3, 1],[3, 2],[3, 3],[3, 4],[3, 5],[3, 6],
          [4, 1],[4, 2],[4, 3],[4, 4],[4, 5],[4, 6],
          [5, 1],[5, 2],[5, 3],[5, 4],[5, 5],[5, 6],
          [6, 1],[6, 2],[6, 3],[6, 4],[6, 5],[6, 6]]

    island.add_thing(charlie, start)
    charlie.performance = 50 #prevents AIMA from resetting the performance to 0 when adding the agent to the environment

    island.add_thing(treasure1A, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure1B, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure1C, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(reusable1, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(reusable2, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(reusable3, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure2A, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure2B, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure2C, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(dispos1, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(dispos2, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(dispos3, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(wall, xy.pop(randint(0, len(xy)) - 1))
    island.add_thing(wall2, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(wall3, xy.pop(randint(0, len(xy) - 1)))

    island.run(50) #run the program for 50 iterations
    return charlie.performance #for studying several runs
                                          
main()

Agent Location: [1, 1]
Agent Tools: []

\ 1 2 3 4 5 6 / 
1 h T - H - X 1 
2 t - - - H H 2 
3 - h - - X - 3 
4 - t T - - - 4 
5 - - - - t X 5 
6 - - T - - h 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 50
Hunter: Grabbed Disposable Tool at [1, 1].
Agent Location: [1, 1]
Agent Tools: ['h']

\ 1 2 3 4 5 6 / 
1 - T - H - X 1 
2 t - - - H H 2 
3 - h - - X - 3 
4 - t T - - - 4 
5 - - - - t X 5 
6 - - T - - h 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 50
Hunter: Moved Right to [1, 2].
Agent Location: [1, 2]
Agent Tools: ['h']

\ 1 2 3 4 5 6 / 
1 - T - H - X 1 
2 t - - - H H 2 
3 - h - - X - 3 
4 - t T - - - 4 
5 - - - - t X 5 
6 - - T - - h 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 49
Hunter: Moved Right to [1, 3].
Agent Location: [1, 3]
Agent Tools: ['h']

\ 1 2 3 4 5 6 / 
1 - T - H - X 1 
2 t - - - H H 2 
3 - h - - X - 3 
4 - t T - - - 4 
5 - - - - t X 5 
6 - - T - - h 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 48
Hunter: Moved Right to [1, 4].
Agent Location: [1, 4]
Agent Tools: ['h']

\ 1 2 3 4 5 6 / 
1 -

204

Next cell executes 500 runs of 50 iterations each of the previous agent and environment. The average agent performance of the runs is computed (takes less than a minute to run but its not inmediate please be patient). The agent location for these runs is also randomized like the location of the objects. Stdout is supressed to avoid excessive printing.

In [31]:
%%capture
total_performance = 0
for i in range (1, 500):
    run_result = main(start=[randint(1, 6), randint(1, 6)]);
    total_performance += run_result
average_performance = total_performance / 500

Next cell prints the average agent performance from the last experiment

In [32]:
average_performance

61.782

# SIMPLE-REFLEX AGENT in PARTIALLY OBSERVABLE ENVIRONMENT #

This model works exactly the same as the last one except for one difference: the percept on each step of a run is not the whole environment anymore. In the percept function of the environment there is a getLocations function which retrieves the location of the cells adjacent to the agent's current position (up to 8 when its not in the border of the grid). These locations then are used to construct the array of percepts which is now limited to adjacent cells.

<b>Next cell runs a simple reflex agent program in a partially observable environment. Output depicts the behavior of the agent for all the steps in a 50 iteration run with 3 items of each kind placed randomly in the environment. The starting position of the agent for this run is \[1, 1\]<b></b></b>

In [33]:
#Import the libraries used for constructing this agent and environment
from agents import *
from random import *
import numpy as np

#Create the classes of the objects placed in the enviroments
class Treasure1(Thing):
    pass

class Treasure2(Thing):
    pass

class DisposTool(Thing):
    pass

class ReuseTool(Thing):
    pass

class Wall(Thing):
    pass



#Create a partially observable class for the environment
class PartialIsland(Environment):
    def __init__(self, width=7, height=7): #Default shape of the environment is square 7x7
        super(PartialIsland, self).__init__()

        self.width = width
        self.height = height
    
    def percept(self, agent): #returns a list of objects that the given agent can sense
        self.percepts = []
        
        locations = self.getPartialLocations(agent.location) #coordinates used to fetch the objects perceivable by the agent

        for locus in locations:
            things = self.list_things_at(locus)
            for thing in things:
                self.percepts.append(thing)
        return self.percepts
    
    def getPartialLocations(self, agent_locus): #returns a list of coordinates of the cells adjacent to the current agent's position
        here = [agent_locus[0], agent_locus[1]]
        up = [agent_locus[0] - 1, agent_locus[1]]
        upright = [agent_locus[0] - 1, agent_locus[1] + 1]
        right = [agent_locus[0], agent_locus[1] + 1]
        downright = [agent_locus[0] + 1, agent_locus[1] + 1]
        down = [agent_locus[0] + 1, agent_locus[1]]
        downleft = [agent_locus[0] + 1, agent_locus[1] - 1]
        left = [agent_locus[0], agent_locus[1] - 1]
        upleft = [agent_locus[0] - 1, agent_locus[1] - 1]
        
        locations = [here, up, upright, right, downright, down, downleft, left, upleft]
        return locations
    
            
    def execute_action(self, agent, action): #this function is called when running the environment after the agent's program returns an action
        '''changes the state of the environment based on what the agent does.'''
        
        if action == 'moveRandom':
            direction = randint(1, 4)
            print("SELECTED ACTION: Random")
            if direction == 1:
                action = 'moveRight'
            elif direction == 2:
                action = 'moveLeft'
            elif direction == 3:
                action = 'moveUp'
            elif direction == 4:
                action = 'moveDown'
                '''This module handles the case when the agent percieved only walls or empty spaces in the cells adjacent to it. The agent in this
                case decides to prepare a random movement action. Another case of this module being called is when the agent is able to percieve
                a treasure but it does not hold a tool that can unlock it. This scenario is effectively the same as ignoring the treasure.'''

        if action == 'moveRight':
            if agent.location[1] < 6:
                walls = self.list_things_at([agent.location[0], agent.location[1]+1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveRight()
                    print("SELECTED ACTION: Right")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move right. First a check is performed to see if the agent decided to move out
                of bounds, in this case the agent's location does not change in the environment and 5 points are deducted from the performance for the
                agent's intent. If the move was in-bounds then it is checked wether or not the destination cell contains a wall; if this is the case the
                agent's decision to move is changed to standing still, otherwise the agent moves and its location is updated in the env successfully'''
                        
        elif action == 'moveLeft':
            if agent.location[1] > 1:
                walls = self.list_things_at([agent.location[0], agent.location[1]-1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveLeft()
                    print("SELECTED ACTION: Left")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move left. It follows the same logic as the previous module'''
        
        elif action == 'moveUp':
            if agent.location[0] > 1:
                walls = self.list_things_at([agent.location[0] - 1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveUp()
                    print("SELECTED ACTION: Up")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move up. It follows the same logic as the previous module'''
                
        elif action == 'moveDown':
            if agent.location[0] < 6:
                walls = self.list_things_at([agent.location[0] + 1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveDown()
                    print("SELECTED ACTION: Down")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move down. It follows the same logic as the previous module'''
                
        elif action == "Greuse":
            items = self.list_things_at(agent.location, tclass=ReuseTool)
            if len(items) != 0:
                if agent.greuse(items[0]):
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    '''This module handles the case when the agent decides to grab a reusable tool. First a confirmation that a reusable tool object is
                    indeed in the current agent's location is performed. Upon success in the check, the agent picks up the tool, then it is deleted from
                    the env (internally from the array of things in the env) and finally the graphic representation of the env is modified accordingly'''
        
        elif action == "Gdispos":
            agent.gdispos()
            items = self.list_things_at(agent.location, tclass=DisposTool)
            if len(items) != 0:
                if agent.gdispos(items[0]):
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    '''This module handles the case when the agent decides to grab a disposable tool. It follows the same logic as the previous module'''
        
        elif action == "GTreasure1":
            items = self.list_things_at(agent.location, tclass=Treasure1)
            if len(items) != 0:
                if agent.gTreasure1(items[0]): 
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    '''This module handles the case when the agent decides to grab a type 1 treasure. It follows the same logic as the previous module'''
        
        elif action == "GTreasure2":
            items = self.list_things_at(agent.location, tclass=Treasure2)
            if len(items) != 0:
                if agent.gTreasure2(items[0]):
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    '''This module handles the case when the agent decides to grab a type 2 treasure. It follows the same logic as the previous module'''
        
        elif action == "NoOp":
            pass
        
        #Report the modified environment along with agent status
        print("NEW AGENT'S PERFORMANCE: " + str(agent.performance))
        print("NEW ENVIRONMENT STATE")
        print("Agent location: " + str(agent.location))
        print("Agent tools: " + str(agent.holding))
        print('\n'.join([''.join(['{:3}'.format(item) for item in row]) 
              for row in self.matrix]))
    
    def run(self, steps=50): #AIMA function overriden to create a graphical representation of the environment and report starting status
        "Run the Environment for given number of time steps."
        print("SIMPLE REFLEX AGENT in PARTIALLY OBSERVABLE ENVIRONMENT")
        print("<STARTING>")
        print("Agent location: " + str(self.things[0].location))
        print("Agent tools: " + str(self.things[0].holding))
        
        self.matrix = templateEnv(size = 6) #create an empty matrix for representing the environment graphically
        self.matrix = fillEnv(self.matrix, self.things) #fill the matrix with objects in the environment
        print("Agent performance: " + str(self.things[0].performance))
        
        for step in range(steps):
            if self.is_done():
                print("\nFINAL AGENT's PERFORMANCE: " + str(self.things[0].performance)) #print the performance after last step of the run
                return
            print("\n<STEP" + str(step + 1) + ">") #print step number starting from 1
            self.step()
        
        print("\nFINAL AGENT's PERFORMANCE: " + str(self.things[0].performance)) #print the performance after last step of the run
        
    def step(self): #AIMA function overriden to access agent status in each step of a run and print the performance
        """Run the environment for one time step. If the
        actions and exogenous changes are independent, this method will
        do.  If there are interactions between them, you'll need to
        override this method."""
        if not self.is_done():
            actions = []
            for agent in self.agents:
                if agent.alive:
                    actions.append(agent.program(agent, self.percept(agent)))
                else:
                    actions.append("")
                
            #Print a graphical matrix representation of what the agent was percieving along with its status at that moment
            print("PERCEPT")
            print("Agent location: " + str(agent.location))
            print("Agent tools: " + str(agent.holding))
            percept_repr = templateEnv(size = 6)
            percept_repr = getUnknownsReflex(percept_repr, agent.location)
            fillEnv(percept_repr, self.percepts)
                
            for (agent, action) in zip(self.agents, actions):
                self.execute_action(agent, action)
            self.exogenous_change()
            
    def is_done(self): #AIMA function overriden to terminate the run if there are no objects of interest left in the environment
        no_edibles = not any(isinstance(thing, Treasure1) or isinstance(thing, DisposTool) or isinstance(thing, ReuseTool) or isinstance(thing, Treasure2) for thing in self.things)
        dead_agents = not any(agent.is_alive() for agent in self.agents)
        return dead_agents or no_edibles
    
    
#Create a simple reflex agent for use in an island environment
class ReflexHunter(Agent):
    def __init__(self, program=None): #default state of the agent
        self.alive = True
        self.bump = False
        self.holding = []
        self.performance = 50
        if program is None:
            def program(percept):
                return eval(input('Percept={}; action? ' .format(percept)))
        assert isinstance(program, collections.Callable)
        self.program = program
        
    #The following 4 methods are actions that the agent executes when it wants to move. Which move of the four is called depends on the calculations
    #made in the reflexProgram() function and on wether or not the action is valid in the current state of the environment as seen in execute_action()
    #in the environment's methods. Performance is reduced by one point in all cases
    def moveRight(self):
        self.performance -= 1
        self.location[1] += 1

    def moveLeft(self):
        self.performance -= 1
        self.location[1] -= 1

    def moveUp(self):
        self.performance -= 1
        self.location[0] -= 1

    def moveDown(self):
        self.performance -= 1
        self.location[0] += 1

    #The following 4 functions are actions that the agent executes when it wants to grab something. They operate in the a similar fashion as the
    #movement methods except that the agent's inventory is also manipulated
    def greuse(self, thing):
        self.performance -= 1
        if isinstance(thing, ReuseTool):
            print("SELECTED ACTION: Greuse")
            self.holding.append('H')
            return True
        return False

    def gdispos(self, thing = None):
        self.performance -= 1
        if isinstance(thing, DisposTool):
            print("SELECTED ACTION: Gdispos")
            self.holding.append('h')
            return True
        return False

    def gTreasure1(self, thing):
        if isinstance(thing, Treasure1):
            print("SELECTED ACTION: GTreas1")
            self.performance += 20
            if 'H' in self.holding == False:
                self.holding.remove('h')
            return True
        return False

    def gTreasure2(self, thing):
        if isinstance(thing, Treasure2):
            print("SELECTED ACTION: GTreas2")
            self.performance += 40
            self.holding.remove('h')
            return True
        return False
    
    #This action is performed when there is a wall at a cell where the agent wants to move to. Mantains the agent's location for the current step
    def NoOp(self):
        print("SELECTED ACTION: NoOp saw a wall")
        return False
    
    
def reflexProgram(agent, percepts):
    '''returns a string dictating the action to be performed by the agent and ultimately what needs to be modified in environment'''
    actionTaken = False
    
    for p in percepts:
        if actionTaken:
            break
            
        in_location = agent.location == p.location #used to check the agent's current location for a grabable object
        
        if isinstance(p, Treasure1) and ('h' in agent.holding or 'H' in agent.holding):
            if in_location:
                actionTaken = True
                return 'GTreasure1'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a type 1 treasure. First check if the agent has a any kind of tool to grab the
                treasure: if not then let this iteration of the loop pass because there is no way to obtain the treasure right now (a new percept is
                checked in the next iteration), if yes check if the treasure is at the same place as the agent, if not move towards it by calling the
                getDirection() function, if yes output the grab action'''
        
        elif isinstance(p, Treasure2) and 'h' in agent.holding:
            if in_location:
                actionTaken = True
                return 'GTreasure2'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a type 2 treasure. It works in the same fashion as the last module with the
                difference being that only disposable tools on inventory are considered in the first condition check and not any kind of tool as in
                the previous module'''
            
        elif isinstance(p, DisposTool):
            if in_location:
                actionTaken = True
                return 'Gdispos'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a disposable tool. I simply checks if the tool is in the current location of the
                agent to pick it up or otherwise move towards it calling the getDirection() function'''
            
        elif isinstance(p, ReuseTool):
            if in_location:
                actionTaken = True
                return 'Greuse'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a reusable tool. Works the same as the last module'''
    
    if not actionTaken:
        return 'moveRandom'
    
def getDirection(origin, goal):
    '''Decides an action that will move the agent one cell closer (in manhatan distance) to an object it wants to pick up'''
    if origin[1] < goal[1]:
        return 'moveRight'
    elif origin[1] > goal[1]:
        return 'moveLeft'
    elif origin [0] > goal[0]:
        return 'moveUp'
    else:
        return 'moveDown'
    
def templateEnv(size):
    '''Creates an empty matrix to represent an enviroment or agent's percept'''
    matrix = np.array([['-' for i in range (0, size + 2)] for j in range (0, size + 2)])
    
    #Make the matrix pretty with outer rol and col aesthetics
    for i in range(0, size + 2):
        for j in range(0, size + 2):
            if (i == 0 and j == 0) or (i == size + 1 and j == size + 1):
                matrix[i][j] = '\\'

            elif (i == 0 and j == size + 1) or (i == size + 1 and j == 0):
                matrix[i][j] = '/'

            elif i == 0 and (j != 0 or j!= size + 1):
                matrix[i][j] = str(j)
            
            elif i == size + 1 and (j != 0 or j!= size + 1):
                matrix[i][j] = str(j)

            elif (i != 0 or i != size + 1) and j == 0:
                matrix[i][j] = str(i)
            
            elif (i != 0 or i != size + 1) and j == size + 1:
                matrix[i][j] = str(i)
    return matrix

def fillEnv(matrix, things):
    '''fills an empty matrix with objects in the environment at their proper locations'''
    
    for thing in things:
        if isinstance(thing, Treasure1):
            matrix[thing.location[0]][thing.location[1]] = 'T'
        elif isinstance(thing, Treasure2):
            matrix[thing.location[0]][thing.location[1]] = 't'
        elif isinstance(thing, DisposTool):
            matrix[thing.location[0]][thing.location[1]] = 'h'
        elif isinstance(thing, ReuseTool):
            matrix[thing.location[0]][thing.location[1]] = 'H'
        elif isinstance(thing, Wall):
            matrix[thing.location[0]][thing.location[1]] = 'X'
        
    print('\n'.join([''.join(['{:3}'.format(item) for item in row]) 
           for row in matrix]))
    return matrix
    
def getUnknownsReflex(matrix, reference):
    '''modifies the graphical representation of the agent's percept to reflect the cells it can't sense (for partial env)'''
    for i in range(1, len(matrix) - 1):
            for j in range(1, len(matrix) - 1):
                
                if [i, j] != reference:
                    if [i+1, j] != reference and [i-1, j] != reference:
                        if [i, j+1] != reference and [i, j-1] != reference:
                            if [i+1, j+1] != reference and [i-1, j-1] != reference:
                                if [i+1, j-1] != reference and [i-1, j+1] != reference:
                                    matrix[i][j] = '?'
    return matrix


def main(start=[1,1]):
    '''Run the simple reflex hunter in the partial treasure island environment'''
    partial_island = PartialIsland()
    mike = ReflexHunter(reflexProgram)

    #Create 3 objects of each kind to place on the environment
    treasure1A = Treasure1()
    treasure1B = Treasure1()
    treasure1C = Treasure1()
    treasure2A = Treasure2()
    treasure2B = Treasure2()
    treasure2C = Treasure2()
    dispos1 = DisposTool()
    dispos2 = DisposTool()
    dispos3 = DisposTool()
    reusable1 = ReuseTool()
    reusable2 = ReuseTool()
    reusable3 = ReuseTool()
    wall = Wall()
    wall2 = Wall()
    wall3 = Wall()

    partial_island.add_thing(mike, start)
    mike.performance = 50 #prevents AIMA from resetting the performance to 0 when adding the agent to the environment
    
    #add things randomly preventing placing more than one item in a single cell in the environment
    xy = [[1, 1],[1, 2],[1, 3],[1, 4],[1, 5],[1, 6],
          [2, 1],[2, 2],[2, 3],[2, 4],[2, 5],[2, 6],
          [3, 1],[3, 2],[3, 3],[3, 4],[3, 5],[3, 6],
          [4, 1],[4, 2],[4, 3],[4, 4],[4, 5],[4, 6],
          [5, 1],[5, 2],[5, 3],[5, 4],[5, 5],[5, 6],
          [6, 1],[6, 2],[6, 3],[6, 4],[6, 5],[6, 6]]

    partial_island.add_thing(treasure1A, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure1B, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure1C, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(reusable1, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(reusable2, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(reusable3, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure2A, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure2B, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure2C, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(dispos1, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(dispos2, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(dispos3, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(wall, xy.pop(randint(0, len(xy)) - 1))
    partial_island.add_thing(wall2, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(wall3, xy.pop(randint(0, len(xy) - 1)))

    partial_island.run(50) #run the program for 50 iterations
    return mike.performance #for studying several runs
                                          
main()

SIMPLE REFLEX AGENT in PARTIALLY OBSERVABLE ENVIRONMENT
<STARTING>
Agent location: [1, 1]
Agent tools: []
\  1  2  3  4  5  6  /  
1  -  T  -  t  -  T  1  
2  -  X  -  X  H  h  2  
3  -  X  h  h  -  -  3  
4  H  -  t  -  -  t  4  
5  -  -  -  -  T  -  5  
6  -  -  -  H  -  -  6  
/  1  2  3  4  5  6  \  
Agent performance: 50

<STEP1>
PERCEPT
Agent location: [1, 1]
Agent tools: []
\  1  2  3  4  5  6  /  
1  -  T  ?  ?  ?  ?  1  
2  -  X  ?  ?  ?  ?  2  
3  ?  ?  ?  ?  ?  ?  3  
4  ?  ?  ?  ?  ?  ?  4  
5  ?  ?  ?  ?  ?  ?  5  
6  ?  ?  ?  ?  ?  ?  6  
/  1  2  3  4  5  6  \  
SELECTED ACTION: Random
SELECTED ACTION: Down
NEW AGENT'S PERFORMANCE: 49
NEW ENVIRONMENT STATE
Agent location: [2, 1]
Agent tools: []
\  1  2  3  4  5  6  /  
1  -  T  -  t  -  T  1  
2  -  X  -  X  H  h  2  
3  -  X  h  h  -  -  3  
4  H  -  t  -  -  t  4  
5  -  -  -  -  T  -  5  
6  -  -  -  H  -  -  6  
/  1  2  3  4  5  6  \  

<STEP2>
PERCEPT
Agent location: [2, 1]
Agent tools: []
\  1  2  3  4  5  6  /  


38

Next cell executes 500 runs of 50 iterations each of the previous agent and environment. The average agent performance of the runs is computed (takes less than a minute to run but its not inmediate please be patient). The agent location for these runs is also randomized like the location of the objects. Stdout is supressed to avoid excessive printing.

In [34]:
%%capture
total_performance = 0
for i in range (1, 500):
    run_result = main(start=[randint(1, 6), randint(1, 6)]);
    total_performance += run_result
average_performance = total_performance / 500

Next cell prints the average agent performance from the last experiment

In [35]:
average_performance

80.446

# MODEL-BASED REFLEX AGENT in FULLY OBSERVABLE ENVIRONMENT #

This model works very similar to the first one (simple-reflex fully observable), but the main change lies in the agent having a built-in memory array for storing past percepts. This difference will be much more significant in the partially observable environment because here the whole grid is always available anyways, making this agent's memory not an actual advantage for this environment.

<b>Next cell runs a model-based reflex agent program in a fully observable environment. Output depicts the behavior of the agent for all the steps in a 50 iteration run with 3 items of each kind placed randomly in the environment. The starting position of the agent for this run is \[1, 1\]<b></b></b>

In [36]:
from agents import *
from random import *


class Treasure1(Thing):
    pass


class Treasure2(Thing):
    pass


class DisposTool(Thing):
    pass


class ReuseTool(Thing):
    pass


class Island(Environment):
    def __init__(self, width=7, height=7):
        super(Island, self).__init__()

        self.width = width
        self.height = height
        # Sets iteration start and end (no walls).
        # self.x_start, self.y_start = (0, 0)
        # self.x_end, self.y_end = (self.width, self.height)

    def percept(self, agent):
        '''prints & return a list of things that are in our agent's location'''
        percepts = []
        locations = []
        # locations = self.getLocations(agent.location)
        # TODO: implement 'getLocations function
        for i in range(1,7):
            for j in range(1,7):
                locations.append([i,j])

        for location in locations:
            things = self.list_things_at(location)
            for thing in things:
                percepts.append(thing)
        return percepts

    def execute_action(self, agent, action):
        '''changes the state of the environment based on what the agent does.'''
        # TODO: check for walls and boundaries
        if action == 'moveRandom':
            direction = randint(1, 4)
            print("Hunter: Moved random")
            if direction == 1:
                action = 'moveRight'
            elif direction == 2:
                action = 'moveLeft'
            elif direction == 3:
                action = 'moveUp'
            elif direction == 4:
                action = 'moveDown'

        if action == 'moveRight':
            if agent.location[1] < 6:
                walls = self.list_things_at([agent.location[0], agent.location[1]+1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveRight()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == 'moveLeft':
            if agent.location[1] > 1:
                walls = self.list_things_at([agent.location[0], agent.location[1] - 1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveLeft()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == 'moveUp':
            if agent.location[0] > 1:
                walls = self.list_things_at([agent.location[0]-1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveUp()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == 'moveDown':
            if agent.location[0] < 6:
                walls = self.list_things_at([agent.location[0] + 1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveDown()
                agent.performance -= 1
            else:
                agent.performance -= 5
        elif action == "Greuse":
            items = self.list_things_at(agent.location, tclass=ReuseTool)
            if len(items) != 0:
                if agent.greuse(items[0]):  #
                    self.delete_thing(items[0])  #
                    agent.holding.append('H')
        elif action == "Gdispos":
            agent.gdispos()
            items = self.list_things_at(agent.location, tclass=DisposTool)
            if len(items) != 0:
                if agent.gdispos(items[0]):  #
                    self.delete_thing(items[0])  # D
                    agent.holding.append('h')
        elif action == "GTreasure1":
            items = self.list_things_at(agent.location, tclass=Treasure1)
            if len(items) != 0:
                if agent.gTreasure1(items[0]):  # Grab Treasure 1
                    # TODO: add to performance
                    agent.performance += 20
                    self.delete_thing(items[0])  # Delete it from the Island after.
        elif action == "GTreasure2":
            items = self.list_things_at(agent.location, tclass=Treasure2)
            if len(items) != 0:
                if agent.gTreasure2(items[0]):  # Grab Treasure2
                    agent.performance += 40
                    agent.holding.remove('h')
                    self.delete_thing(items[0])  # Delete it from the Island after.
        elif action == "NoOp":
            pass

    def is_done(self):
        '''By default, we're done when we can't find a live agent,
        but to prevent killing our cute dog, we will or it with when there is no more food or water'''
        no_edibles = not any(isinstance(thing, Treasure1) or isinstance(thing, DisposTool) or isinstance(thing, ReuseTool) or isinstance(thing, Treasure2) for thing in self.things)
        dead_agents = not any(agent.is_alive() for agent in self.agents)
        return dead_agents or no_edibles

    def step(self): #AIMA function overriden to access agent status in each step of a run and print the performance
        """Run the environment for one time step. If the
        actions and exogenous changes are independent, this method will
        do.  If there are interactions between them, you'll need to
        override this method."""
        if not self.is_done():
            actions = []
            for agent in self.agents:
                if agent.alive:
                    actions.append(agent.program(self.percept(agent), agent))
                else:
                    actions.append("")
                
            for (agent, action) in zip(self.agents, actions):
                self.execute_action(agent, action)
            self.exogenous_change()

class ReflexHunter(Agent):
    def __init__(self, program=None):  # default state of the agent
        self.alive = True
        self.bump = False
        self.type = 'simple_reflex'
        self.holding = []
        self.performance = 50
        if program is None:
            def program(percept):
                return eval(input('Percept={}; action? '.format(percept)))
        # assert isinstance(program, collections.Callable)
        self.program = program

    def moveRight(self):
        self.location[1] += 1
        print("Hunter: Moved Right to {}.".format(self.location))

    def moveLeft(self):
        self.location[1] -= 1
        print("Hunter: Moved Left to {}.".format(self.location))

    def moveUp(self):
        self.location[0] -= 1
        print("Hunter: Moved Up to {}.".format(self.location))

    def moveDown(self):
        self.location[0] += 1
        print("Hunter: Moved Down to {}.".format(self.location))

    def greuse(self, thing):
        '''returns True upon success or False otherwise'''
        if isinstance(thing, ReuseTool):
            print("Hunter: Grabbed Reusable Tool at {}.".format(self.location))
            return True
        return False

    def gdispos(self, thing = None):
        ''' returns True upon success or False otherwise'''
        if isinstance(thing, DisposTool):
            print("Hunter: Grabbed Disposable Tool at {}.".format(self.location))
            return True
        return False

    def gTreasure1(self, thing):
        ''' returns True upon success or False otherwise'''
        if isinstance(thing, Treasure1):
            print("Hunter: Grabbed Treasure1 at {}.".format(self.location))
            return True
        return False

    def gTreasure2(self, thing):
        ''' returns True upon success or False otherwise'''
        if isinstance(thing, Treasure2):
            print("Hunter: Grabbed Treasure2 at {}.".format(self.location))
            return True
        return False
    
class ModelBasedHunter(ReflexHunter):
    def __init__(self, program=None):  # default state of the agent
        self.alive = True
        self.bump = False
        self.holding = []
        self.performance = 50
        self.type = "modelbased"
        self.seen = [[1,2], [1,3]]
        if program is None:
            def program(percept):
                return eval(input('Percept={}; action? '.format(percept)))
        # assert isinstance(program, collections.Callable)
        self.program = program
    
def program(percepts, agent):
    '''Returns an action based on it's percepts'''
    print("Agent Location: " + str(agent.location))
    print("Agent Tools: " + str(agent.holding))
    printMatrix(percepts)
    print("")
    print("Agent performance: " + str(agent.performance))
    if agent.type == 'modelbased':
        print("INTERNAL STATE:")
        print(agent.seen)
    actionTaken = False
    for p in percepts:
        # Grab actions for when agent is in same location
        if actionTaken:
            break
        in_location = agent.location == p.location
        if isinstance(p, Treasure1) and inInventory('H', agent):
            if in_location:
                actionTaken = True
                return 'GTreasure1'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
        elif isinstance(p, Treasure2) and inInventory('h', agent):
            if in_location:
                actionTaken = True
                return 'GTreasure2'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
        elif isinstance(p, DisposTool):
            if in_location:
                actionTaken = True
                return 'Gdispos'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
        elif isinstance(p, ReuseTool):
            if in_location:
                actionTaken = True
                return 'Greuse'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo

    if not actionTaken:
        return 'moveRandom'

def printMatrix(percepts):
    for i in range(0, 8):
        print("")
        for j in range(0, 8):
            if (i == 0 and j == 0) or (i == 7 and j == 7):
                print("\\", end=" ")

            elif (i == 0 and j == 7) or (i == 7 and j == 0):
                print("/", end=" ")

            elif (i == 0 or i == 7):
                print("".join(str(j)), end=" ")

            elif (j == 0 or j == 7):
                print("".join(str(i)), end=" ")

            else:
                #Check percepts, print if matches location
                printed = False
                for p in percepts:
                    if p.location == [i,j]:
                        if isinstance(p, Treasure1):
                            printed = True
                            print("T", end=" ")
                        elif isinstance(p, Treasure2):
                            printed = True
                            print("t", end=" ")
                        elif isinstance(p, DisposTool):
                            printed = True
                            print("h", end=" ")
                        elif isinstance(p, ReuseTool):
                            printed = True
                            print("H", end=" ")
                        elif isinstance(p, Wall):
                            printed = True
                            print("X", end=" ")
                if not printed:
                    print("-", end=" ")


def inInventory(tool, agent):
    for hold in agent.holding:
        if hold == tool:
            return True
    return False

def getDirection(origin, goal):
    if origin[1] < goal[1]:
        return 'moveRight'
    elif origin[1] > goal[1]:
        return 'moveLeft'
    elif origin[0] > goal[0]:
        return 'moveUp'
    else:
        return 'moveDown'
    
def main(start = [1, 1]):
    '''Run the simple reflex hunter in the partial treasure island environment'''
    island = Island()
    charlie = ModelBasedHunter(program)

    #Create 3 objects of each kind to place on the environment
    treasure1A = Treasure1()
    treasure1B = Treasure1()
    treasure1C = Treasure1()
    treasure2A = Treasure2()
    treasure2B = Treasure2()
    treasure2C = Treasure2()
    dispos1 = DisposTool()
    dispos2 = DisposTool()
    dispos3 = DisposTool()
    reusable1 = ReuseTool()
    reusable2 = ReuseTool()
    reusable3 = ReuseTool()
    wall = Wall()
    wall2 = Wall()
    wall3 = Wall()
    
    #add things randomly preventing placing more than one item in a single cell in the environment
    xy = [[1, 1],[1, 2],[1, 3],[1, 4],[1, 5],[1, 6],
          [2, 1],[2, 2],[2, 3],[2, 4],[2, 5],[2, 6],
          [3, 1],[3, 2],[3, 3],[3, 4],[3, 5],[3, 6],
          [4, 1],[4, 2],[4, 3],[4, 4],[4, 5],[4, 6],
          [5, 1],[5, 2],[5, 3],[5, 4],[5, 5],[5, 6],
          [6, 1],[6, 2],[6, 3],[6, 4],[6, 5],[6, 6]]

    island.add_thing(charlie, start)
    charlie.performance = 50 #prevents AIMA from resetting the performance to 0 when adding the agent to the environment

    island.add_thing(treasure1A, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure1B, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure1C, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(reusable1, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(reusable2, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(reusable3, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure2A, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure2B, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(treasure2C, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(dispos1, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(dispos2, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(dispos3, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(wall, xy.pop(randint(0, len(xy)) - 1))
    island.add_thing(wall2, xy.pop(randint(0, len(xy) - 1)))
    island.add_thing(wall3, xy.pop(randint(0, len(xy) - 1)))

    island.run(50) #run the program for 50 iterations
    return charlie.performance #for studying several runs
                                          
main()

Agent Location: [1, 1]
Agent Tools: []

\ 1 2 3 4 5 6 / 
1 - - - t - - 1 
2 t - - H - T 2 
3 - T H H - h 3 
4 - - - t - h 4 
5 - - h - T X 5 
6 X - - - - X 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 50
INTERNAL STATE:
[[1, 2], [1, 3]]
Hunter: Moved Right to [1, 2].
Agent Location: [1, 2]
Agent Tools: []

\ 1 2 3 4 5 6 / 
1 - - - t - - 1 
2 t - - H - T 2 
3 - T H H - h 3 
4 - - - t - h 4 
5 - - h - T X 5 
6 X - - - - X 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 49
INTERNAL STATE:
[[1, 2], [1, 3]]
Hunter: Moved Right to [1, 3].
Agent Location: [1, 3]
Agent Tools: []

\ 1 2 3 4 5 6 / 
1 - - - t - - 1 
2 t - - H - T 2 
3 - T H H - h 3 
4 - - - t - h 4 
5 - - h - T X 5 
6 X - - - - X 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 48
INTERNAL STATE:
[[1, 2], [1, 3]]
Hunter: Moved Right to [1, 4].
Agent Location: [1, 4]
Agent Tools: []

\ 1 2 3 4 5 6 / 
1 - - - t - - 1 
2 t - - H - T 2 
3 - T H H - h 3 
4 - - - t - h 4 
5 - - h - T X 5 
6 X - - - - X 6 
/ 1 2 3 4 5 6 \ 
Agent performance: 47
INTERNAL STATE:


171

Next cell executes 500 runs of 50 iterations each of the previous agent and environment. The average agent performance of the runs is computed (takes less than a minute to run but its not inmediate please be patient). The agent location for these runs is also randomized like the location of the objects. Stdout is supressed to avoid excessive printing.

In [37]:
%%capture
total_performance = 0
for i in range (1, 500):
    run_result = main(start=[randint(1, 6), randint(1, 6)]);
    total_performance += run_result
average_performance = total_performance / 500

Next cell prints the average agent performance from the last experiment

In [38]:
average_performance

66.96

# MODEL-BASED REFLEX AGENT in PARTIALLY OBSERVABLE ENVIRONMENT #

This time around the agent takes full advantage of its memory capabilities and stores useful objects it was able to sense at some point during the run. During the first step of a run the whole percept is appended to the memory. For the remaining of the steps it is checked wether or not a new array of percepts actually contains new objects not previously seen before. If that is the case they are added to the memory. It is worth mentioning that this particular model only remembers relevant objects in memory and not actually empty spaces, so empty spaces that are not adjacent to the agent will return to a "?" status from a "-" status (empty).

<b>Next cell runs a model-based reflex agent program in a partially observable environment. Output depicts the behavior of the agent for all the steps in a 50 iteration run with 3 items of each kind placed randomly in the environment. The starting position of the agent for this run is \[1, 1\]<b></b></b>

In [39]:
#Import the libraries used for constructing this agent and environment
from agents import *
from random import *
import numpy as np

#Create the classes of the objects placed in the enviroments
class Treasure1(Thing):
    pass

class Treasure2(Thing):
    pass

class DisposTool(Thing):
    pass

class ReuseTool(Thing):
    pass

class Wall(Thing):
    pass



#Create a partially observable class for the environment
class PartialIsland(Environment):
    def __init__(self, width=7, height=7): #Default shape of the environment is square 7x7
        super(PartialIsland, self).__init__()

        self.width = width
        self.height = height
    
    def percept(self, agent): #returns a list of objects that the given agent can sense
        self.percepts = []
        
        locations = self.getPartialLocations(agent.location) #coordinates used to fetch the objects perceivable by the agent

        for locus in locations:
            things = self.list_things_at(locus)
            for thing in things:
                self.percepts.append(thing)
        return self.percepts
    
    def getPartialLocations(self, agent_locus): #returns a list of coordinates of the cells adjacent to the current agent's position
        here = [agent_locus[0], agent_locus[1]]
        up = [agent_locus[0] - 1, agent_locus[1]]
        upright = [agent_locus[0] - 1, agent_locus[1] + 1]
        right = [agent_locus[0], agent_locus[1] + 1]
        downright = [agent_locus[0] + 1, agent_locus[1] + 1]
        down = [agent_locus[0] + 1, agent_locus[1]]
        downleft = [agent_locus[0] + 1, agent_locus[1] - 1]
        left = [agent_locus[0], agent_locus[1] - 1]
        upleft = [agent_locus[0] - 1, agent_locus[1] - 1]
        
        locations = [here, up, upright, right, downright, down, downleft, left, upleft]
        return locations
    
            
    def execute_action(self, agent, action): #this function is called when running the environment after the agent's program returns an action
        '''changes the state of the environment based on what the agent does.'''
        
        if action == 'moveRandom':
            direction = randint(1, 4)
            print("SELECTED ACTION: Random")
            if direction == 1:
                action = 'moveRight'
            elif direction == 2:
                action = 'moveLeft'
            elif direction == 3:
                action = 'moveUp'
            elif direction == 4:
                action = 'moveDown'
                '''This module handles the case when the agent percieved only walls or empty spaces in the cells adjacent to it. The agent in this
                case decides to prepare a random movement action. Another case of this module being called is when the agent is able to percieve
                a treasure but it does not hold a tool that can unlock it. This scenario is effectively the same as ignoring the treasure.'''

        if action == 'moveRight':
            if agent.location[1] < 6:
                walls = self.list_things_at([agent.location[0], agent.location[1]+1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveRight()
                    print("SELECTED ACTION: Right")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move right. First a check is performed to see if the agent decided to move out
                of bounds, in this case the agent's location does not change in the environment and 5 points are deducted from the performance for the
                agent's intent. If the move was in-bounds then it is checked wether or not the destination cell contains a wall; if this is the case the
                agent's decision to move is changed to standing still, otherwise the agent moves and its location is updated in the env successfully'''
                        
        elif action == 'moveLeft':
            if agent.location[1] > 1:
                walls = self.list_things_at([agent.location[0], agent.location[1]-1], tclass=Wall)
                if len(walls) == 0:
                    agent.moveLeft()
                    print("SELECTED ACTION: Left")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move left. It follows the same logic as the previous module'''
        
        elif action == 'moveUp':
            if agent.location[0] > 1:
                walls = self.list_things_at([agent.location[0] - 1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveUp()
                    print("SELECTED ACTION: Up")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move up. It follows the same logic as the previous module'''
                
        elif action == 'moveDown':
            if agent.location[0] < 6:
                walls = self.list_things_at([agent.location[0] + 1, agent.location[1]], tclass=Wall)
                if len(walls) == 0:
                    agent.moveDown()
                    print("SELECTED ACTION: Down")
                else:
                    agent.NoOp()
            else:
                print("SELECTED ACTION: canceled, tried to move out of bounds")
                agent.performance -= 5
                '''This module handles the case when the agent decided to move down. It follows the same logic as the previous module'''
                
        elif action == "Greuse":
            items = self.list_things_at(agent.location, tclass=ReuseTool)
            if len(items) != 0:
                if agent.greuse(items[0]):
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    agent.memories.remove(items[0])
                    '''This module handles the case when the agent decides to grab a reusable tool. First a confirmation that a reusable tool object is
                    indeed in the current agent's location is performed. Upon success in the check, the agent picks up the tool, then it is deleted from
                    the env (internally from the array of things in the env) and finally the graphic representation of the env is modified accordingly. 
                    The tool existence in the agent's memory is also erased'''
        
        elif action == "Gdispos":
            agent.gdispos()
            items = self.list_things_at(agent.location, tclass=DisposTool)
            if len(items) != 0:
                if agent.gdispos(items[0]):
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    agent.memories.remove(items[0])
                    '''This module handles the case when the agent decides to grab a disposable tool. It follows the same logic as the previous module'''
        
        elif action == "GTreasure1":
            items = self.list_things_at(agent.location, tclass=Treasure1)
            if len(items) != 0:
                if agent.gTreasure1(items[0]): 
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    agent.memories.remove(items[0])
                    '''This module handles the case when the agent decides to grab a type 1 treasure. It follows the same logic as the previous module'''
        
        elif action == "GTreasure2":
            items = self.list_things_at(agent.location, tclass=Treasure2)
            if len(items) != 0:
                if agent.gTreasure2(items[0]):
                    self.delete_thing(items[0])
                    self.matrix[items[0].location[0]][items[0].location[1]] = '-'
                    agent.memories.remove(items[0])
                    '''This module handles the case when the agent decides to grab a type 2 treasure. It follows the same logic as the previous module'''
        
        elif action == "NoOp":
            pass
        
        #Report the modified environment along with agent status
        print("NEW AGENT'S PERFORMANCE: " + str(agent.performance))
        print("NEW ENVIRONMENT STATE")
        print("Agent location: " + str(agent.location))
        print("Agent tools: " + str(agent.holding))
        print('\n'.join([''.join(['{:3}'.format(item) for item in row]) 
              for row in self.matrix]))
    
    def run(self, steps=50): #AIMA function overriden to create a graphical representation of the environment and report starting status
        "Run the Environment for given number of time steps."
        print("MODEL-BASED REFLEX AGENT in PARTIALLY OBSERVABLE ENVIRONMENT")
        print("<STARTING>")
        print("Agent location: " + str(self.things[0].location))
        print("Agent tools: " + str(self.things[0].holding))
        
        self.matrix = templateEnv(size = 6) #create an empty matrix for representing the environment graphically
        self.matrix = fillEnv(self.matrix, self.things) #fill the matrix with objects in the environment
        print("Agent performance: " + str(self.things[0].performance))
        
        for step in range(steps):
            if self.is_done():
                print("\nFINAL AGENT's PERFORMANCE: " + str(self.things[0].performance)) #print the performance after last step of the run
                return
            print("\n<STEP" + str(step + 1) + ">") #print step number starting from 1
            self.step()
        
        print("\nFINAL AGENT's PERFORMANCE: " + str(self.things[0].performance)) #print the performance after last step of the run
        
    def step(self): #AIMA function overriden to access agent status in each step of a run and print the performance
        """Run the environment for one time step. If the
        actions and exogenous changes are independent, this method will
        do.  If there are interactions between them, you'll need to
        override this method."""
        if not self.is_done():
            actions = []
            for agent in self.agents:
                if agent.alive:
                    actions.append(agent.program(agent, self.percept(agent)))
                else:
                    actions.append("")
                
            #Print a graphical matrix representation of what the agent was percieving along with its status at that moment
            print("PERCEPT")
            print("Agent location: " + str(agent.location))
            print("Agent tools: " + str(agent.holding))
            percept_repr = templateEnv(size = 6)
            getUnknownsModel(percept_repr, agent.location, agent.memories)
                
            for (agent, action) in zip(self.agents, actions):
                self.execute_action(agent, action)
            self.exogenous_change()
            
    def is_done(self): #AIMA function overriden to terminate the run if there are no objects of interest left in the environment
        no_edibles = not any(isinstance(thing, Treasure1) or isinstance(thing, DisposTool) or isinstance(thing, ReuseTool) or isinstance(thing, Treasure2) for thing in self.things)
        dead_agents = not any(agent.is_alive() for agent in self.agents)
        return dead_agents or no_edibles
    
    
#Create a simple reflex agent for use in an island environment
class ModelHunter(Agent):
    def __init__(self, program=None): #default state of the agent
        self.alive = True
        self.bump = False
        self.holding = []
        self.performance = 50
        self.memories = [] #this is exclusive to the model-based agent, stores useful percepts it has seen in previous steps of a run
        if program is None:
            def program(percept):
                return eval(input('Percept={}; action? ' .format(percept)))
        assert isinstance(program, collections.Callable)
        self.program = program
        
    def remember(self, percepts):
        '''This function is called from the program of the agent, first check if there is something on the agent's memory. If its the first step of the
        run this will be true and the whole current percept is appended to the memory. If its not the first step of the run then the agent needs to
        check if the current percept contain useful information that he previously didn't know. New useful percepts are appended to the memory.'''
        if len(self.memories) == 0:
            self.memories = percepts
        else:
            temp = percepts
            for thing in self.memories:
                for p in percepts:
                    if p is thing:
                        temp.remove(p)
                    else:
                        pass
            if len(temp) > 0:
                for p in temp:
                    self.memories.append(p)
            else:
                pass
        
    #The following 4 methods are actions that the agent executes when it wants to move. Which move of the four is called depends on the calculations
    #made in the reflexProgram() function and on wether or not the action is valid in the current state of the environment as seen in execute_action()
    #in the environment's methods. Performance is reduced by one point in all cases
    def moveRight(self):
        self.performance -= 1
        self.location[1] += 1

    def moveLeft(self):
        self.performance -= 1
        self.location[1] -= 1

    def moveUp(self):
        self.performance -= 1
        self.location[0] -= 1

    def moveDown(self):
        self.performance -= 1
        self.location[0] += 1

    #The following 4 functions are actions that the agent executes when it wants to grab something. They operate in the a similar fashion as the
    #movement methods except that the agent's inventory is also manipulated
    def greuse(self, thing):
        self.performance -= 1
        if isinstance(thing, ReuseTool):
            print("SELECTED ACTION: Greuse")
            self.holding.append('H')
            return True
        return False

    def gdispos(self, thing = None):
        self.performance -= 1
        if isinstance(thing, DisposTool):
            print("SELECTED ACTION: Gdispos")
            self.holding.append('h')
            return True
        return False

    def gTreasure1(self, thing):
        if isinstance(thing, Treasure1):
            print("SELECTED ACTION: GTreas1")
            self.performance += 20
            if 'H' in self.holding == False:
                self.holding.remove('h')
            return True
        return False

    def gTreasure2(self, thing):
        if isinstance(thing, Treasure2):
            print("SELECTED ACTION: GTreas2")
            self.performance += 40
            self.holding.remove('h')
            return True
        return False
    
    #This action is performed when there is a wall at a cell where the agent wants to move to. Mantains the agent's location for the current step
    def NoOp(self):
        print("SELECTED ACTION: NoOp saw a wall")
        return False
    
    
def modelProgram(agent, percepts):
    '''returns a string dictating the action to be performed by the agent and ultimately what needs to be modified in environment'''
    agent.remember(percepts) #call this method to append useful current percepts to the agent's memory
    
    actionTaken = False
    
    for p in agent.memories: #works with the whole agent's memory and not just the current percept
        if actionTaken:
            break
            
        in_location = agent.location == p.location #used to check the agent's current location for a grabable object
        
        if isinstance(p, Treasure1) and ('h' in agent.holding or 'H' in agent.holding):
            if in_location:
                actionTaken = True
                return 'GTreasure1'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a type 1 treasure. First check if the agent has a any kind of tool to grab the
                treasure: if not then let this iteration of the loop pass because there is no way to obtain the treasure right now (a new percept is
                checked in the next iteration), if yes check if the treasure is at the same place as the agent, if not move towards it by calling the
                getDirection() function, if yes output the grab action'''
        
        elif isinstance(p, Treasure2) and 'h' in agent.holding:
            if in_location:
                actionTaken = True
                return 'GTreasure2'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a type 2 treasure. It works in the same fashion as the last module with the
                difference being that only disposable tools on inventory are considered in the first condition check and not any kind of tool as in
                the previous module'''
            
        elif isinstance(p, DisposTool):
            if in_location:
                actionTaken = True
                return 'Gdispos'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a disposable tool. I simply checks if the tool is in the current location of the
                agent to pick it up or otherwise move towards it calling the getDirection() function'''
            
        elif isinstance(p, ReuseTool):
            if in_location:
                actionTaken = True
                return 'Greuse'
            else:
                moveTo = getDirection(agent.location, p.location)
                actionTaken = True
                return moveTo
                '''This module handles when the agent is able to sense a reusable tool. Works the same as the last module'''
    
    if not actionTaken:
        return 'moveRandom'
    
def getDirection(origin, goal):
    '''Decides an action that will move the agent one cell closer (in manhatan distance) to an object it wants to pick up'''
    if origin[1] < goal[1]:
        return 'moveRight'
    elif origin[1] > goal[1]:
        return 'moveLeft'
    elif origin [0] > goal[0]:
        return 'moveUp'
    else:
        return 'moveDown'
    
def templateEnv(size):
    '''Creates an empty matrix to represent an enviroment or agent's percept'''
    matrix = np.array([['-' for i in range (0, size + 2)] for j in range (0, size + 2)])
    
    #Make the matrix pretty with outer rol and col aesthetics
    for i in range(0, size + 2):
        for j in range(0, size + 2):
            if (i == 0 and j == 0) or (i == size + 1 and j == size + 1):
                matrix[i][j] = '\\'

            elif (i == 0 and j == size + 1) or (i == size + 1 and j == 0):
                matrix[i][j] = '/'

            elif i == 0 and (j != 0 or j!= size + 1):
                matrix[i][j] = str(j)
            
            elif i == size + 1 and (j != 0 or j!= size + 1):
                matrix[i][j] = str(j)

            elif (i != 0 or i != size + 1) and j == 0:
                matrix[i][j] = str(i)
            
            elif (i != 0 or i != size + 1) and j == size + 1:
                matrix[i][j] = str(i)
    return matrix

def fillEnv(matrix, things):
    '''fills an empty matrix with objects in the environment at their proper locations'''
    
    for thing in things:
        if isinstance(thing, Treasure1):
            matrix[thing.location[0]][thing.location[1]] = 'T'
        elif isinstance(thing, Treasure2):
            matrix[thing.location[0]][thing.location[1]] = 't'
        elif isinstance(thing, DisposTool):
            matrix[thing.location[0]][thing.location[1]] = 'h'
        elif isinstance(thing, ReuseTool):
            matrix[thing.location[0]][thing.location[1]] = 'H'
        elif isinstance(thing, Wall):
            matrix[thing.location[0]][thing.location[1]] = 'X'
        
    print('\n'.join([''.join(['{:3}'.format(item) for item in row]) 
           for row in matrix]))
    return matrix
    
def getUnknownsModel(matrix, reference, memories):
    '''modifies the graphical representation of the agent's percept to reflect the cells it can't sense and the cells it has seen
    in previous steps that contain relevant objects (for partial env)'''
    for thing in memories:
        if isinstance(thing, Treasure1):
            matrix[thing.location[0]][thing.location[1]] = 'T'
        elif isinstance(thing, Treasure2):
            matrix[thing.location[0]][thing.location[1]] = 't'
        elif isinstance(thing, DisposTool):
            matrix[thing.location[0]][thing.location[1]] = 'h'
        elif isinstance(thing, ReuseTool):
            matrix[thing.location[0]][thing.location[1]] = 'H'
        elif isinstance(thing, Wall):
            matrix[thing.location[0]][thing.location[1]] = 'X'
    
    for i in range(1, len(matrix) - 1):
            for j in range(1, len(matrix) - 1):
                
                if matrix[i, j] == '-':
                    if [i, j] != reference:
                        if [i+1, j] != reference and [i-1, j] != reference:
                            if [i, j+1] != reference and [i, j-1] != reference:
                                if [i+1, j+1] != reference and [i-1, j-1] != reference:
                                    if [i+1, j-1] != reference and [i-1, j+1] != reference:
                                        matrix[i][j] = '?'
                                        
    print('\n'.join([''.join(['{:3}'.format(item) for item in row]) 
          for row in matrix]))

def main(start=[1, 1]):
    '''Run the model-based reflex hunter in the partial treasure island environment'''
    partial_island = PartialIsland()
    mike = ModelHunter(modelProgram)

    #Create 3 objects of each kind to place on the environment
    treasure1A = Treasure1()
    treasure1B = Treasure1()
    treasure1C = Treasure1()
    treasure2A = Treasure2()
    treasure2B = Treasure2()
    treasure2C = Treasure2()
    dispos1 = DisposTool()
    dispos2 = DisposTool()
    dispos3 = DisposTool()
    reusable1 = ReuseTool()
    reusable2 = ReuseTool()
    reusable3 = ReuseTool()
    wall = Wall()
    wall2 = Wall()
    wall3 = Wall()

    partial_island.add_thing(mike, start)
    mike.performance = 50 #prevents AIMA from resetting the performance to 0 when adding the agent to the environment
    
    #add things randomly preventing placing more than one item in a single cell in the environment
    xy = [[1, 1],[1, 2],[1, 3],[1, 4],[1, 5],[1, 6],
          [2, 1],[2, 2],[2, 3],[2, 4],[2, 5],[2, 6],
          [3, 1],[3, 2],[3, 3],[3, 4],[3, 5],[3, 6],
          [4, 1],[4, 2],[4, 3],[4, 4],[4, 5],[4, 6],
          [5, 1],[5, 2],[5, 3],[5, 4],[5, 5],[5, 6],
          [6, 1],[6, 2],[6, 3],[6, 4],[6, 5],[6, 6]]

    partial_island.add_thing(treasure1A, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure1B, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure1C, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(reusable1, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(reusable2, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(reusable3, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure2A, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure2B, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(treasure2C, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(dispos1, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(dispos2, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(dispos3, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(wall, xy.pop(randint(0, len(xy)) - 1))
    partial_island.add_thing(wall2, xy.pop(randint(0, len(xy) - 1)))
    partial_island.add_thing(wall3, xy.pop(randint(0, len(xy) - 1)))

    partial_island.run(50) #run the program for 50 iterations
    return mike.performance #for studying several runs
                                          
main()

MODEL-BASED REFLEX AGENT in PARTIALLY OBSERVABLE ENVIRONMENT
<STARTING>
Agent location: [1, 1]
Agent tools: []
\  1  2  3  4  5  6  /  
1  -  -  -  T  X  -  1  
2  -  -  T  -  -  X  2  
3  t  -  H  -  -  h  3  
4  -  h  -  -  h  t  4  
5  -  -  T  t  -  -  5  
6  -  H  -  -  H  X  6  
/  1  2  3  4  5  6  \  
Agent performance: 50

<STEP1>
PERCEPT
Agent location: [1, 1]
Agent tools: []
\  1  2  3  4  5  6  /  
1  -  -  ?  ?  ?  ?  1  
2  -  -  ?  ?  ?  ?  2  
3  ?  ?  ?  ?  ?  ?  3  
4  ?  ?  ?  ?  ?  ?  4  
5  ?  ?  ?  ?  ?  ?  5  
6  ?  ?  ?  ?  ?  ?  6  
/  1  2  3  4  5  6  \  
SELECTED ACTION: Random
SELECTED ACTION: Down
NEW AGENT'S PERFORMANCE: 49
NEW ENVIRONMENT STATE
Agent location: [2, 1]
Agent tools: []
\  1  2  3  4  5  6  /  
1  -  -  -  T  X  -  1  
2  -  -  T  -  -  X  2  
3  t  -  H  -  -  h  3  
4  -  h  -  -  h  t  4  
5  -  -  T  t  -  -  5  
6  -  H  -  -  H  X  6  
/  1  2  3  4  5  6  \  

<STEP2>
PERCEPT
Agent location: [2, 1]
Agent tools: []
\  1  2  3  4  5  6 

67

Next cell executes 500 runs of 50 iterations each of the previous agent and environment. The average agent performance of the runs is computed (takes less than a minute to run but its not inmediate please be patient). The agent location for these runs is also randomized like the location of the objects. Stdout is supressed to avoid excessive printing.

In [40]:
%%capture
total_performance = 0
for i in range (1, 500):
    run_result = main(start=[randint(1, 6), randint(1, 6)]);
    total_performance += run_result
average_performance = total_performance / 500

Next cell prints the average agent performance from the last experiment

In [41]:
average_performance

85.458

# CONCLUSIONS #

<b> Who behaves better, the simple reflex agent or the model based reflex agent? </b>

They perform fairly similar, depending on the type of the environment and the configuration of the thing in it. In the fully observable environment they perform exactly the same way, while on most configurations for the partially observable environment the model-based agent is able to perform much better due to its ability to avoid series of random movement to revisit objects it has already seen at some point. Average performance over 500 random partially observable environments favours the model-based agent consistenly by about 10 points over the simple-reflex agent (85-90 over 75-80).


To observate better the performance Simple-Reflex (SR) and Model-Based (MB) on both enviroments Fully-Observable (FO) and Partial-Observable (PO) we calculate an Average Perfomance Measure that consist on 500 runs, the results are prompted in the following table:


<table>
    <tr>
        <td><b>Agent:</b> </td>
        <td>SR-FO </td>
        <td>SR-PO</td>
        <td>MB-FO</td>
         <td>MB-PO</td>
      

   </tr>
   <tr>
       <td><b>Agent´s Average Perfomance measure:</b> </td>
       <td>61.782</td>
       <td>80.446</td>
       <td>66.96</td>
       <td>85.458</td>
       
   </tr>
        
</table>


<b> How do different types of environments affect different types of agents? </b>

In a fully observable environment, both agents perform the same. Having every Thing Object in the percepts lists for every iteration makes it redundant to save an internal memory, thus making the model-based agent no different than the simple reflex.

However, on the partially-observable environment, the model-based agent is able to perform much better, as every step taken allows it to get to know the island better and remember the location of desirable objects even if it has to come back later for them.

The difference on the agent performance resides on what they are able to perceive. Since the simple reflex has no memory, in the partial environment it will be moving mostly at random until it stumbles upon a desirable object.

<b> Do you think the agents are behaving rationally? </b>

Yes, they are making the best decision on the information that they have. Except when they face walls which we found to be time consuming to be dealt with in an intelligent way. If a desirable object is behind a wall, the agents will be crashing into it everytime until they run out of iterations. The rules for going past a wall were not part of the scope for this assignment. One quick solution which minimizes performance loss would be to call the "NoOp" action to avoid losing performance even if the agent doesn't move away from the wall.

<b> What did you have to program differently for agents to work in different environments? </b>

The main feature we had to program out of the specifications of the assignment to get results in different environments was the option to move random when no action can be taken based on percepts. This way, the agent won't be stuck on the same position if it can't grab a treasure at its current location. Between the environments themselves the percept function was what changed mainly and the inclusion (or not) of an array for storing past percepts was the main difference between agents.
